### PR TITLE
Add authentication filter to listener filters chain.

### PR DESCRIPTION
**Pull request by diemtvu**
Authentication filter is Istio specific filter to perform authentication logic as specified in policy. If policy is not provided (or empty), it does nothing.

Issue: https://github.com/istio/istio/issues/4337

### CODE REVIEW

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 1]
```diff
@@ -155,18 +161,48 @@ func ConvertPolicyToJwtConfig(policy *authn.Policy) *jwtfilter.JwtAuthentication
 	return ret
 }
 
+// ConvertPolicyToAuthNFilterConfig returns authn filter config corresponding for the input policy.
+func ConvertPolicyToAuthNFilterConfig(policy *authn.Policy) *authn_filter.FilterConfig {
+	if policy == nil {
+		return &authn_filter.FilterConfig{Policy: &authn.Policy{}}
+	}
+	filterConfig := &authn_filter.FilterConfig{
+		Policy: &authn.Policy{
```
Comment (qiwzhang) [inline pilot/pkg/networking/plugin/authn/authentication.go]: why don't we just do:

     Policy:  policy

Comment (diemtvu) [reply pilot/pkg/networking/plugin/authn/authentication.go]: I need to delete the targets part of the policy, and much as I know about go, policy is a pointer so it actually modify the original one (which break the pilot test, when we reuse policy in for different tests).

Anyway, I make a change here to clone the message and delete the unwanted part.

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 2]
```diff
@@ -155,14 +162,56 @@ func ConvertPolicyToJwtConfig(policy *authn.Policy) *jwtfilter.JwtAuthentication
 	return ret
 }
 
+// ConvertPolicyToAuthNFilterConfig returns authn filter config corresponding for the input policy.
```
Comment (ZackButcher) [inline pilot/pkg/networking/plugin/authn/authentication.go]: returns _an_ authentication filter config corresponding _to_ the input policy

File: pilot/pkg/networking/plugin/authn/authentication_test.go [hunk 3]
```diff
@@ -534,3 +535,222 @@ func TestBuildJwtFilter(t *testing.T) {
...
+					{
+						Params: &authn.PeerAuthenticationMethod_Jwt{
+							Jwt: &authn.Jwt{
+								JwksUri: "http://abc.com",
+							},
+						},
+					},
+				},
+			},
+			expected: &http_conn.HttpFilter{
```
Comment (ZackButcher) [inline pilot/pkg/networking/plugin/authn/authentication_test.go]: Can you clean this up, e.g. by using the original type inline and wrapping it in a call to convert it to a struct? Struct protos are really hard to read.

File: pilot/pkg/proxy/envoy/v1/authentication.go [hunk 4]
```diff
@@ -46,7 +39,25 @@ func buildJwtFilter(policy *authn.Policy) *HTTPFilter {
...
 	return &HTTPFilter{
 		Type:   decoder,
-		Name:   jwtFilterName,
+		Name:   authn_plugin.JwtFilterName,
+		Config: config,
+	}
+}
+
+// buildAuthnFilter returns a authN filter for the policy.
+func buildAuthnFilter(policy *authn.Policy) *HTTPFilter {
```
Comment (ZackButcher) [inline pilot/pkg/proxy/envoy/v1/authentication.go]: Atm we don't want to change the v1 code at all. All new features are going in to v2 only. This code should live in v2 and v1 should remain unchanged (cc @costinm).

File: pilot/pkg/networking/plugin/authn/authentication.go (hunk 2 above)
Comment (diemtvu) [reply pilot/pkg/networking/plugin/authn/authentication.go]: Done

File: pilot/pkg/networking/plugin/authn/authentication_test.go (hunk 3 above)
Comment (diemtvu) [reply pilot/pkg/networking/plugin/authn/authentication_test.go]: I change the test to convert config back to proto and compare with golden proto value. Hope it's easier to read

(will clean up other test in another if this is a reasonable way to do)

File: pilot/pkg/proxy/envoy/v1/authentication.go (hunk 4 above)
Comment (diemtvu) [reply pilot/pkg/proxy/envoy/v1/authentication.go]: For the record, the filter will be hooked up with v2 listener via plugin. I want to wait for some plugin refactor (e.g [this](https://github.com/istio/istio/pull/4718)) before doing so.

**Pull request merged**